### ISSUE TITLE
Icons mangled (home icon jumps to top of screen, wastebin superimposed) on refresh

### ISSUE THREAD

**Issue by CottonEaster** [linuxmint/nemo#1654]
```
 * Nemo version (nemo --version): 3.6.4
 * Is issue with desktop or windowed nemo? Desktop
 * Distribution - (Mint 17.2, Arch, Fedora 25, etc...) Mint 17.3 Sylvia
 * Graphics hardware *and* driver used: Intel Mobile 965; 2.1 Mesa 17.0.7
 * 32 or 64 bit: 64
 ```

**Issue**
After the upgrade to Mint 18.3, F5 pushes my 'home' icon to the top of screen and pushes my wastebin under another icon. Toggling 'align to grid' has the same affect.

**Steps to reproduce**

As above.

**Expected behaviour**

No jumping and no superposition.

**Other information**
Before:
![before](https://user-images.githubusercontent.com/9998846/33315857-cc50ab5a-d429-11e7-8ead-fb4def7c1c9c.png)
After:
![before](https://user-images.githubusercontent.com/9998846/33315863-ceea60cc-d429-11e7-8e5c-db0222551b5a.png)

EDIT: Thank you for Mint 18.3, which very largely is working very well!

FURTHER EDIT: Thic icon re-arranging is happening every time I boot.

ALSO: cf. [https://github.com/linuxmint/nemo/issues/358](url)

**Comment by CottonEaster** [comment]
I should add this: I've discovered that the problem affects only icons added to the desktop from Settings -> Desktop.

**Comment by mtwebster** [comment]
Are those the same two images by accident?  Or I'm blind..   I remember I thought I fixed this, or something related, I'll check again.

**Comment by CottonEaster** [comment]
> Are those the same two images by accident?

Yes. _Sorry!_ I've fixed it now.

> I remember I thought I fixed this, or something related

Well, I had the same problem - and reported it - some months ago on Mint 18.2. Some other people had it too. But then it was fixed. But now, on 18.3, it has reappeared.

**Comment by CottonEaster** [comment]
A video showing the full extent - which is fuller than I realised - of the problem: [dancing icons](https://app.box.com/s/54bd4skbsfx1pa9kimqavi8i3kpku9cu).

**Comment by mtwebster** [comment]
Can you put that somewhere else? I can't view it.  Not previewable, now download option (even with an account for Box).

**Comment by CottonEaster** [comment]
Sorry. It should be downloadable now. I don't think the link has changed but [here it is again just in case](https://app.box.com/s/54bd4skbsfx1pa9kimqavi8i3kpku9cu).

**Comment by mtwebster** [comment]
Can you post the contents of your `~/.config/nemo/desktop-metadata` as well as the output from `dconf dump /org/nemo/`

This is also reported here: https://github.com/linuxmint/nemo/issues/1674

I cannot reproduce at this time, so additional info is needed.

**Comment by CottonEaster** [comment]
deskto-metadata: [pastebin](https://pastebin.com/zfQPY5iu). (There's a lot of old stuff in there, to do with drives that haven't been in the computer for ages.)

dconf: [pastebin](https://pastebin.com/bScJSy5H).

Thanks.

(That other report is not by me, by the way.)

**Comment by CottonEaster** [comment]
I wonder whether this problem might owe to permissions of the icon files concerned.

**Comment by mtwebster** [comment]
I've found the commit that introduced it, working on ironing out a solution currently.  I think I've actually already fixed, just looking at a few other quirks I noticed.

**Comment by CottonEaster** [comment]
Great. Might I suggest that the problem is sufficiently serious to warrant the release of the fix before Mint 19 appears? For, the problem occurs every boot and will be throwing for, especially, new users.

**Comment by mtwebster** [comment]
Mind trying these packages before I get a new release out:

https://www.dropbox.com/s/zan64tyruumqfwa/nemo_3.6.5_amd64.deb?dl=0
https://www.dropbox.com/s/wyvpayruuhl60du/nemo-data_3.6.5_all.deb?dl=0
https://www.dropbox.com/s/cxzbs8w3sl7xmjw/libnemo-extension1_3.6.5_amd64.deb?dl=0
https://www.dropbox.com/s/o4llf1uy3soapiu/gir1.2-nemo-3.0_3.6.5_amd64.deb?dl=0
https://www.dropbox.com/s/c8x7vdm3x7kj80h/nemo-dbg_3.6.5_amd64.deb?dl=0
https://www.dropbox.com/s/3ua9l1ymfbv16m1/libnemo-extension-dev_3.6.5_amd64.deb?dl=0

Download all packages, open a terminal in the folder you downloaded to then:

```
sudo dpkg -i *nemo*.deb
```

If something goes wrong, or you want to revert, you can let mintupdate install the repo version.

You'll need to restart nemo and nemo-desktop:
```
nemo --quit
nemo-desktop --quit
```

nemo-desktop should restart automatically, if it doesn't, just log out and back in, or run `nemo-desktop` with no options.

Thanks

**Comment by CottonEaster** [comment]
@mtwebster 

I appreciate this and I am happy to test.

The results of the test are as follows.

After manual rearrange:
![image](https://user-images.githubusercontent.com/9998846/34720181-b3f4ffb6-f535-11e7-8748-a95a77538fc6.png)

After F5, and on boot:
![image](https://user-images.githubusercontent.com/9998846/34720144-951e5db2-f535-11e7-96b3-1b72bc326a44.png)

Relevant packages: 

![image](https://user-images.githubusercontent.com/9998846/34720115-81851f2a-f535-11e7-974c-519d548916ff.png)

In short: the problem is not solved yet.

**Comment by mtwebster** [comment]
I just noticed you had auto arrange off. Is this still the case?  If so, can you test with it on?  Any reason you have it off?  Is it just trouble finding good layout spacing?  Thanks

**Comment by CottonEaster** [comment]
I have auto-arrange off because I want my icons as they were and not like this:

![image](https://user-images.githubusercontent.com/9998846/34721517-ee390316-f53a-11e7-908c-45fa09a4f88e.png)

(By the way: some sort of visual cue - turning the pointer to a padlock? - might stop people repeatedly trying to move locked icons.)

But, yes: *with auto-arrange on (and the new nemo packages installed), F5* (I haven't tested boot yet) *does not rearrange the icons*. They stay were they were (which is not where I want them).

*Another way the icon positions survive F5* (with, again, the new packages installed): *auto-arrange off, but align-to-grid on*. Screenshot:

![image](https://user-images.githubusercontent.com/9998846/34721630-583e6e72-f53b-11e7-92fa-e1128d72856a.png)


And the icon position with the grid on prompts this observation: when the grid, and auto-arrange, are off, F5 puts the system (Home, Bin) icons where they would be were auto-arrange on . .

**Comment by mtwebster** [comment]
Sorry I meant keep-aligned not auto arrange - I understand auto arrange not being usable in your case. I’ll check some additional things for when keep align is off.

**Comment by CottonEaster** [comment]
I should add the following. I discover that, (even) when align-to-grid is off, my Home icon gets displaced (by boot and by F5).

I'd thought that, because the Bin had stopped moving around, all was well when alignment is disabled; but it is not. However, I may have reverted back to the older version of Nemo; I'm not sure (sorry), but you can judge from this:

![image](https://user-images.githubusercontent.com/9998846/34783453-884549d6-f623-11e7-9ee3-32890a0a5c63.png)

Also: when I created a new desktop file just now, it appeared smack-bang atop the Home icon.

**Comment by mtwebster** [comment]
Updated packages to ttry (you had indeed 'updated' back to the distro packages, but no matter):

https://www.dropbox.com/s/ba9c4csi875hax2/nemo_3.6.5_amd64.deb?dl=0
https://www.dropbox.com/s/wvkuojzl39tyx1u/nemo-data_3.6.5_all.deb?dl=0
https://www.dropbox.com/s/orwqjok2n5g575b/gir1.2-nemo-3.0_3.6.5_amd64.deb?dl=0
https://www.dropbox.com/s/rxpce9uctw1jwnp/nemo-dbg_3.6.5_amd64.deb?dl=0
https://www.dropbox.com/s/334mq1rcdsmo927/libnemo-extension1_3.6.5_amd64.deb?dl=0
https://www.dropbox.com/s/ecoaf5g5ni1qg3k/libnemo-extension-dev_3.6.5_amd64.deb?dl=0

A word on the "align to grid" setting:  I'm going to try very hard to switch to strictly grid position metadata for the coming next release.  Meaning, I won't store desktop icons by pixel position (which was carried over from the 'old' desktop for reasons, at the time, of compatibility and re-use of existing code.) - instead, they'll be stored as grid positions.  This will, ironically, simplify a lot of code, much of which is contributing to some of these positioning issues, and hopefully improve robustness (resolutions changing, re-sizing, etc.. will be a bit more orderly, rather than forcing one to redo their layout.)  

A consequence of this is there will no longer be a non-keep-aligned mode.  All positioning will be snapped to a grid.  Hopefully the layouts themselves will be improved in the process (eliminating the invisible borders that can't be positioned onto currently,) making this not that big a deal.

I plan on keeping the 'legacy' desktop mode, which is currently a non-visible dconf setting.

Please let me know how this set of packages works.  One note, when keep-aligned is off, there's no interference checking.  If I unplug a usb device, then move an icon to that device's icons' former spot, they *will* overlap.  Unfortunately implementing this properly is too involved at this point, considering it will be going away.

**Comment by CottonEaster** [comment]
**Results with the new packages, with auto-arrange disabled**
========================================

**1. With align-to-grid disabled**

_Boot_: icons were were I left them.

**F5**: for some seconds - I am unsure how many, but at least six (see below) - the desktop disappeared completely. I opened a log viewer, full-screen, to investigate. I found nothing. Upon returnning to the desktop, all was as it should be. Repeated pressings of F5 worked as they should. Perhaps the delayed desktop owes to nemo doing a one-time fix of problems.

**2. With align-to-grid enabled**

_Boot_: icons were were I (/the grid) had left them. (Note: this reboot occured after testing with align-to-grid disabled.)

_F5_: All as it should be, though seemingly the more granular, full-screen-covering grid has yet to arrive (I wasn't sure whether that was meant to have arrived yet), thus leaving me with an icon layout that I don't really want.

All in all: much better.

PS: Immediately after installing the packages, I had trouble restarting nemo, perhaps because I ran the quit commands as root, which Mint didn't seem to like. So/anyhow, I rebooted.

**Comment by CottonEaster** [comment]
Also, and still with align enabled, I plugged in a usb stick. An icon for it appeared on the desktop in a place that did not overlap any icon nor indeed did it overlap my conky. I ejected it (sadly there are still _two_ context menu commands to do that) and reinserted it in a different usb port and, er, nothing appeared on the desktop, though non-window nemo saw the stick. I then repeated the procedure (eject, reinsert) and, again, no icon.

**Comment by CottonEaster** [comment]
On resuming from suspend and trying to move a row of icons, I could not move the icons at all. Now admittedly - and cf. a suggestion I made earlier in this thread - one might think that is because I was not trying to move them far enough (for, they will move only to the next grid position, and give no indication when that does not occur, save for not budging). However, it seemed, for a good twenty seconds or so, that, no, they would not move at all. To have a usable, stable, system - albeit one in which icons have a mind of their own - I am afraid that, for the moment, I am going back to the repository versions.

**Comment by mtwebster** [comment]
That sounds related to resume, not the desktop itself.  Was the rest of your system responsive during this time (other than the mouse itself)?

**Comment by CottonEaster** [comment]
I think the rest of the system was responsive, but I can't rule out something else - to do with resume and not to do with nemo - being responsible.

I do have a script that runs on resume, and another on wifi connection. Those scripts are meant to work around various problems, none of which, perhaps, owe to Mint itself as against bugs in and infelicities with: (i) third-party software; (ii) the hardware of my computer. I could test the new nemo packages on another computer, though admittedly that computer has the same scripts. Or I could try disabling the scripts temporarily. Does everything work alright at your end, on a fresh install? One of my computers (both my Mint PCs are underpowered laptops) just about has the muscle to run a clean version of Mint in a VM, but you will understand, I hope, that there is only so much testing that it is practical for me to do.

**Comment by CottonEaster** [comment]
Michael, perhaps the problem was indeed my end. For, I've given those packages another ago and all seems to be well (though I have yet to re-test the insert-remove-insert USB device cycle).

**Comment by CottonEaster** [comment]
There's a tiny bit of icon overlap, even with the new packages and with align enabled:

![image](https://user-images.githubusercontent.com/9998846/34900663-ef6ca2c6-f7f9-11e7-88ff-74e58b5cd968.png)

**Comment by CottonEaster** [comment]
I should report also that, with align enabled and the new packages, my attempts to re-position icons - drag them to a new place - sometimes has strange results. To wit: total overlap (i.e. superposition) and at other times the icon moving to an unoccupied but wrong position. However, the problem is intermittent, and might possibly have something to do with my 'Rainlendar' calendar on my desktop.

**Comment by CottonEaster** [comment]
It might be worth noting the following. On a clean install of Cinnamon, on a new computer, with auto-arrange disabled, but align-to-grid on, and no Conky, an icon for a USB stick (an icon that appears when the stick is mounted) landed slap bang atop another item. Admittedly, I had not yet applied the patches from this page. I'll do some now. Perhaps those patches should be rolled out via the update manager.

**Comment by CottonEaster** [comment]
I've had the problem, severely, on another new install (and on a third computer). The following steps should suffice to reproduce the problem.

1. Install Mint Cinnamon x64 18.3
2. Put some icons on the desktop.
3. Disable 'auto-align'.
4. Insert USB stick.
5. Press F5 a few times.

Step 4 generated a very much superimposed icon. Step 5 - or perhaps it was disabling align-to-grid - resulting in all my icons being atop each other, in one single place on the desktop.

**Comment by cdgoldin** [comment]
I downloaded 
   https://www.dropbox.com/s/ba9c4csi875hax2/nemo_3.6.5_amd64.deb?dl=0
   https://www.dropbox.com/s/wvkuojzl39tyx1u/nemo-data_3.6.5_all.deb?dl=0
   https://www.dropbox.com/s/orwqjok2n5g575b/gir1.2-nemo-3.0_3.6.5_amd64.deb?dl=0
   https://www.dropbox.com/s/rxpce9uctw1jwnp/nemo-dbg_3.6.5_amd64.deb?dl=0
   https://www.dropbox.com/s/334mq1rcdsmo927/libnemo-extension1_3.6.5_amd64.deb?dl=0
   https://www.dropbox.com/s/ecoaf5g5ni1qg3k/libnemo-extension-dev_3.6.5_amd64.deb?dl=0
to a folder, and followed the instructions listed (way) above:
   Download all packages, open a terminal in the folder you downloaded to then:
  sudo dpkg -i *nemo*.deb
  nemo --quit
   nemo-desktop --quit
I encountered the following errors:
cdg@System5 ~/Downloads/NemoMods $ dir
gir1.2-nemo-3.0_3.6.5_amd64.deb        nemo_3.6.5_amd64.deb
libnemo-extension1_3.6.5_amd64.deb     nemo-data_3.6.5_all.deb
libnemo-extension-dev_3.6.5_amd64.deb  nemo-dbg_3.6.5_amd64.deb
cdg@System5 ~/Downloads/NemoMods $ sudo dpkg -i *nemo* .deb
[sudo] password for cdg: 
dpkg: warning: downgrading gir1.2-nemo-3.0 from 3.6.5+sylvia to 3.6.5
(Reading database ... 344789 files and directories currently installed.)
Preparing to unpack gir1.2-nemo-3.0_3.6.5_amd64.deb ...
Unpacking gir1.2-nemo-3.0 (3.6.5) over (3.6.5+sylvia) ...
dpkg: warning: downgrading libnemo-extension1 from 3.6.5+sylvia to 3.6.5
Preparing to unpack libnemo-extension1_3.6.5_amd64.deb ...
Unpacking libnemo-extension1 (3.6.5) over (3.6.5+sylvia) ...
Selecting previously unselected package libnemo-extension-dev.
Preparing to unpack libnemo-extension-dev_3.6.5_amd64.deb ...
Unpacking libnemo-extension-dev (3.6.5) ...
dpkg: warning: downgrading nemo from 3.6.5+sylvia to 3.6.5
Preparing to unpack nemo_3.6.5_amd64.deb ...
Unpacking nemo (3.6.5) over (3.6.5+sylvia) ...
dpkg: warning: downgrading nemo-data from 3.6.5+sylvia to 3.6.5
Preparing to unpack nemo-data_3.6.5_all.deb ...
Unpacking nemo-data (3.6.5) over (3.6.5+sylvia) ...
Selecting previously unselected package nemo-dbg.
Preparing to unpack nemo-dbg_3.6.5_amd64.deb ...
Unpacking nemo-dbg (3.6.5) ...
dpkg: error processing archive .deb (--install):
 cannot access archive: No such file or directory
Setting up libnemo-extension1 (3.6.5) ...
dpkg: dependency problems prevent configuration of libnemo-extension-dev:
 libnemo-extension-dev depends on libglib2.0-dev (>= 2.37.3); however:
  Package libglib2.0-dev is not installed.
 libnemo-extension-dev depends on libgtk-3-dev (>= 3.9.10); however:
  Package libgtk-3-dev is not installed.

dpkg: error processing package libnemo-extension-dev (--install):
 dependency problems - leaving unconfigured
Setting up nemo-data (3.6.5) ...
Setting up gir1.2-nemo-3.0 (3.6.5) ...
Processing triggers for libglib2.0-0:i386 (2.48.2-0ubuntu1) ...
Processing triggers for libglib2.0-0:amd64 (2.48.2-0ubuntu1) ...
Processing triggers for shared-mime-info (1.5-2ubuntu0.1) ...
Unknown media type in type 'all/all'
Unknown media type in type 'all/allfiles'
Processing triggers for hicolor-icon-theme (0.15-0ubuntu1) ...
Setting up nemo (3.6.5) ...
Processing triggers for menu (2.1.47ubuntu1) ...
Processing triggers for desktop-file-utils (0.22+linuxmint1) ...
Setting up nemo-dbg (3.6.5) ...
Processing triggers for libc-bin (2.23-0ubuntu10) ...
Processing triggers for man-db (2.7.5-1) ...
Processing triggers for gnome-menus (3.13.3-6ubuntu3.1) ...
Processing triggers for mime-support (3.59ubuntu1) ...
Errors were encountered while processing:
 .deb
 libnemo-extension-dev

Nevertheless, MOST of the problems reported appear to be solved:
1) The system icons (i.e. computer, home, network, trash and mounted volumes) are no longer overlaid on top of each other upon reboot or log-on.
2) Pressing F5 (which is supposed to refresh the desktop) no longer causes more of them to be overlaid on top of each other, until they all share the single icon position at the top left of the screen.
3) Saving a (new) file to the desktop, or mounting a device, which is supposed to create an icon in the first OPEN position (left-to-right, top-to-bottom), now appears to work correctly.
4) "Align to grid" STILL aligns to a much more compact grid, which is far less readable because it crams the icon descriptions into the icon images. Note that this occurs with "icon size: smaller". I don't know if it occurs with "normal" or "larger", as it is difficult to test without redoing the entire desktop.

**Comment by CottonEaster** [comment]
@cdgoldin 

The only real error in the message you report is the stuff about dependencies. I encountered those errors on one of my machines. The solution, if I recall, is to the following and the re-run the commands you tried:

`sudo apt install libglib2.0-dev libgtk-3-dev`

(though I may not have got the package names exactly right).

**Comment by CottonEaster** [comment]
I was bolder than you: I installed `libglib2.0-dev` and `libgtk-3-dev` (and I have checked just now that I do have those packages on my system and that the command that I gave above for installing them is correct). And then the patched Nemo files installed without errors (the warnings about downgrading are not errors, and are to be expected). So I think I've taken the risk for you.

Here are a few further points.

- By installing patches you are taking a risk anyway.
- With Linux - though admittedly how desirable this is depends on your circumstances - often one learns by breaking and then fixing things. You've got backups, right?
- In case it is not clear: I am not a Mint developer.

**Comment by cdgoldin** [comment]
Thank you. I appreciate your assistance. What I meant was that, if you weren't sure what you did to solve the problem, it would be foolish for me to follow the unsure advice, since I wouldn't know what to do to correct it if it failed. Your subsequent post is more reassuring.
IMO, patches should be thoroughly tested, and NOT constitute a risk, but I guess that hasn't been true of the software industry for many decades.
Yes, I have backups. And I had to restore from them today, after applying the patches, because I was unable to boot afterwards. Not sure if this was the result of the missing dependencies or the subsequent Nemo update resulting in a "broken package". It doesn't seem that either should have corrupted grub or boot sectors.
And, yes, I realize you are not a Mint Developer. I'm curious why the developers haven't addressed this problem yet. I assume they read these same forums.

**Comment by JosephMcc** [comment]
@cdgoldin First off you ran
`cdg@System5 ~/Downloads/NemoMods $ sudo dpkg -i nemo .deb`
That is not the command listed above. It should have been
`sudo dpkg -i *nemo*.deb`
The wildcards are important which is why you had the error at the end about install .deb
Also the two missing `dev` packages are totally safe to install. They would only be an issue if you have some weird ppa installed that contains versions that shouldn't be in MInt. I doubt that's the case since the rest of you system is broken :)

**Comment by cdgoldin** [comment]
Thank you for your comment. Actually, I ran the correct dpkg command (sudo dpkg -i *nemo*.deb), which I cut-and-pasted from the instructions earlier in this thread. For some reason, the forum software has removed the asterisks from my comment above (but not in my .txt file copy of it). So THAT was NOT the cause of the error messages, but rather the fact that the supplied patch references two packages that are not installed in a normal Linux Mint 18.3 system. It would also seem that the patch triggers Update Manager to reinstall the earlier version of Nemo, and THIS results in a broken package and unbootable system, which is not desirable.

**Comment by CottonEaster** [comment]
I have the problem no longer. Good job!

**Comment by ChooChooAl** [comment]
IF this is closed, does that mean that Nemo 3.6.5 is going to have patches applied via the standard repros for all of us other users?    I do not want to add all of these patches manually on all of my systems.
I was going to add a bug to github on this and found it reported on #1674, and that issue is still open!

**Comment by CottonEaster** [comment]
@ChooChooAl 

Apologies: I see that my comment was unclear. Let me clarify. _What solved my problem standard Nemo updates_, not any hack or patch mentioned on this page. So: if your system is up to date, you should not have the problem. Or at least that is the case _if_ the other bug report that you mentioned (#1674) can (and should) be closed . .

I suspect that the Mint developers pay little attention to closed issues. So, if you still have a problem, then it seems to me that you should chime in on that other, open bug report (#1674); and, if that gets you nowhere, perhaps you should open a new bug report.

**Comment by JosephMcc** [comment]
> Who closed this issue? Since the problem has NOT been fixed, it should still be open!

The person who originally opened the issue. It was solved for him, so he closed it. Perfectly valid on his part.

> "Mint developers" pay little attention to ANY issues reported. That's why nothing ever gets fixed.

You don't have the slightest clue how much time developers spend troubleshooting issues that are reported here. You could have at least a minimal amount of respect for people who spend so much of their free time trying to improve Mint and it's tools. If it's so broken for you, feel free to give a hand trying to fix it. This attitude doesn't help anyone.

**Comment by cdgoldin** [comment]
My question ("Who closed this issue) was in respect to the issue I opened, https://bugs.launchpad.net/bugs/1738302 not #1654, which I certainly did NOT close. Since I am receiving emails discussing both (overlapping) issues, it is confusing (to say the least) which issue is being referenced.

You are correct. I have no idea how much time "developers" spend troubleshooting issues, since there is little evidence that they have done anything in respect to the many issues I have reported in the last decade. I do know that the issues reported in #1654 and #1738302 should never have occurred. How can you release untested software? Or how can you claim you have tested software when such a severe flaw exists for everyone who uses Mint 18.3? Respect is earned, not ordained. It is ludicrous to expect USERS to fix bugs in software released by "developers". 

----- Original Message -----

**Comment by JosephMcc** [comment]
I have no idea about the launchpad issues. We really don't use it anymore and I personally don't subscribe to them. As far as I can tell, you've only ever opened two issues on github. One which is not something we control, so not a Linux Mint bug. The other was about this same issue, which is still open. Both of the issues related to the icons have had feedback in them from the person who actually works on them. Trying to get them solved. Just because an issue is reproducible by a user doesn't mean it is by a developer who may have never seen the problem.

To be frank, if you have such an issue with our approach to development, their are a ton of other choices out there.

**Comment by leigh123linux** [comment]
@cdgoldin  So you managed to get this locked as I'm getting sick of reading your drivel.
